### PR TITLE
feat: PoC of auto-generating solc-remaps file (for Etherscan Platform)

### PR DESCRIPTION
**Pull request by shortdoom**
Related: https://github.com/crytic/crytic-compile/issues/543

PoC as it only deals with the most common scenario when relative imports were originally structured around @package-name. For foundry projects it's most common to specify remappings string without the "@" and there are also more messier examples in the wild. I think it should be possible to cover most of them though.

What does it do?

After downloading and compiling code with the use of Etherscan platform it will output `solc_remaps.txt` file with "_guessed_" remappings strings, built out of absolute paths used internally by crytic-compile at first download. Then, it will save the file to the working directory of project (downloaded files) 

How to use?

No additional command is needed. Run crytic-compile as usual with address on the network as the target. After it finishes

`cd` into the root of downloaded project _(crytic-export/etherscan-contracts/0x1-Test)_ and run:

 `crytic-compile contracts/Target.sol --solc-remaps $(cat solc_remaps.txt)`

### CODE REVIEW

**Comment by 0xalpharush** [comment]
What is the reason to guess the remapping instead of saving the ones downloaded from etherscan here in `crytic-export/remappings.txt`? 
https://github.com/crytic/crytic-compile/blob/1e5deb2e06e2d606444b30767af421c9f4a2d55e/crytic_compile/platform/etherscan.py#L185-L190

We can probably automatically load the `remappings.txt` file and not require the user to pass `--solc-remaps` as well

Wdyt?

**Comment by shortdoom** [comment]
Those `remappings` seem to be always empty (`None`) for me, no matter what the target when I was running manual testing. I verified most of this with slapping logging functions here and there through out the execution flow of CryticCompile => Etherscan => solc_standard_json.standalone_compile(). 

as far as I could understood, CompilationUnit actually just uses absolute paths from here to compile on the first run:

https://github.com/crytic/crytic-compile/blob/1b2998e0215179c910b085c66c7f49df3169098f/crytic_compile/compilation_unit.py#L36 

hence why I also generated just absolute paths for the `solc_remaps.txt`. those seem more reliable than relative.

`remappings` variable seems to, for some reason, not be used or at least doesn't generate any usable remapping strings for next compilation

**Comment by elopez** [comment]
Hi @shortdoom! thanks for working on this!

I verified locally and the remapping variable does have proper remappings, is it possible you tested with a contract on etherscan that does not have the metadata? I printed the remappings like the following:

```diff
diff --git a/crytic_compile/platform/etherscan.py b/crytic_compile/platform/etherscan.py
index a8f54b4..cbec350 100644
--- a/crytic_compile/platform/etherscan.py
+++ b/crytic_compile/platform/etherscan.py
@@ -187,6 +187,10 @@ def _handle_multiple_files(
 
     remappings = dict_source_code.get("settings", {}).get("remappings", None)
 
+    print("Remappings:", remappings)
+    print("Directory:", directory)
+    print("Sanitized:", _sanitize_remappings(remappings, directory))
+
     return list(filtered_paths), directory, _sanitize_remappings(remappings, directory)
 
 ```
 
 And fetching an address, I do get remappings:
 ```sh
% ./venv/bin/crytic-compile 0x2a311e451491091d2a1d3c43f4f5744bdb4e773a
Remappings: ['@0x/contracts-utils=/home/cluracan/code/0x-protocol/node_modules/@0x/contracts-utils', '@0x/contracts-erc20=/home/cluracan/code/0x-protocol/contracts/zero-ex/node_modules/@0x/contracts-erc20']
Directory: crytic-export/etherscan-contracts/0x2a311e451491091d2a1d3c43f4f5744bdb4e773a-NativeOrdersFeature
Sanitized: ['@0x/contracts-utils=home/cluracan/code/0x-protocol/node_modules/@0x/contracts-utils/', '@0x/contracts-erc20=home/cluracan/code/0x-protocol/contracts/zero-ex/node_modules/@0x/contracts-erc20/']
INFO:CryticCompile:'solc --standard-json --allow-paths /Users/emilio/git/crytic-compile/crytic-export/etherscan-contracts/0x2a311e451491091d2a1d3c43f4f5744bdb4e773a-NativeOrdersFeature' running
```

**Comment by shortdoom** [comment]
Huh, yes! It seems that targets I was running it against were without the metadata. 

That definitely changes the purpose of this PR, happy to hear how you would want to proceed.

Ps. Nice target address for testing path resolution edge cases :)

**Comment by elopez** [comment]
yeah that's a cool address to test things with, there's other ones on the `ci_etherscan_test.sh` file if you want to try them out :)

by the way, regarding the address you were using to test on Slack (`0x29d2bcf0d70f95ce16697e645e2b76d218d66109`), it seems that one doesn't require any remappings to build; that likely explains why it works while there's no remappings on the metadata.

```sh
emilio@mbpro 0x29d2bcf0d70f95ce16697e645e2b76d218d66109-OxODexPool % crytic-compile contracts/OxODexPool.sol
INFO:CryticCompile:'solc --version' running
INFO:CryticCompile:'solc contracts/OxODexPool.sol --combined-json abi,ast,bin,bin-runtime,srcmap,srcmap-runtime,userdoc,devdoc,hashes --allow-paths .,/Users/emilio/git/crytic-compile/crytic-export/etherscan-contracts/0x29d2bcf0d70f95ce16697e645e2b76d218d66109-OxODexPool/contracts' running
INFO:CryticCompile:Compilation warnings/errors on contracts/OxODexPool.sol:
Warning: SPDX license identifier not provided in source file. Before publishing, consider adding a comment containing "SPDX-License-Identifier: <SPDX-License>" to each source file. Use "SPDX-License-Identifier: UNLICENSED" for non-open-source code. Please see https://spdx.org for more information.
--> contracts/interfaces/IWETH9.sol

Warning: Unnamed return variable can remain unassigned. Add an explicit return with value to all non-reverting code paths or name the variable.
  --> contracts/lib/LSAG.sol:21:18:
   |
21 |         returns (uint256[2] memory)
   |                  ^^^^^^^^^^^^^^^^^

Warning: Unused local variable.
   --> contracts/OxODexPool.sol:337:24:
    |
337 |         (bool success, bytes memory data) = _recipient.call{value: _amount - relayerFee}("");
    |                        ^^^^^^^^^^^^^^^^^

```

On an unrelated note, it seems globbing of files is not working as expected @0xalpharush , I had to patch it as follows to build all the files at the same time. I saw there's a TODO on that area so I'm bringing it to your attention here.

```diff
diff --git a/crytic_compile/crytic_compile.py b/crytic_compile/crytic_compile.py
index 3996150..8f2b896 100644
--- a/crytic_compile/crytic_compile.py
+++ b/crytic_compile/crytic_compile.py
@@ -717,8 +717,8 @@ def compile_all(target: str, **kwargs: str) -> List[CryticCompile]:
         else:
             compilations.append(CryticCompile(target, **kwargs))
     elif os.path.isdir(target):
-        solidity_filenames = glob.glob(os.path.join(target, "*.sol"))
-        vyper_filenames = glob.glob(os.path.join(target, "*.vy"))
+        solidity_filenames = glob.glob(os.path.join(target, "**", "*.sol"), recursive=True)
+        vyper_filenames = glob.glob(os.path.join(target, "**", "*.vy"), recursive=True)
         # Determine if we're using --standard-solc option to
         # aggregate many files into a single compilation.
         if use_solc_standard_json:
```

With that change you can build all the files from `0x29d2bcf0d70f95ce16697e645e2b76d218d66109` at once after the initial download:

```sh
emilio@mbpro 0x29d2bcf0d70f95ce16697e645e2b76d218d66109-OxODexPool % ../../../venv/bin/crytic-compile . --solc-standard-json
INFO:CryticCompile:'solc --version' running
INFO:CryticCompile:'solc --standard-json --allow-paths /Users/emilio/git/crytic-compile/crytic-export/etherscan-contracts/0x29d2bcf0d70f95ce16697e645e2b76d218d66109-OxODexPool' running
WARNING:CryticCompile:Warning: Warning: SPDX license identifier not provided in source file. Before publishing, consider adding a comment containing "SPDX-License-Identifier: <SPDX-License>" to each source file. Use "SPDX-License-Identifier: UNLICENSED" for non-open-source code. Please see https://spdx.org for more information.
--> ./contracts/interfaces/IWETH9.sol


Warning: Warning: Unnamed return variable can remain unassigned. Add an explicit return with value to all non-reverting code paths or name the variable.
  --> ./contracts/lib/LSAG.sol:21:18:
   |
21 |         returns (uint256[2] memory)
   |                  ^^^^^^^^^^^^^^^^^


Warning: Warning: Unused local variable.
   --> ./contracts/OxODexPool.sol:337:24:
    |
337 |         (bool success, bytes memory data) = _recipient.call{value: _amount - relayerFee}("");
    |                        ^^^^^^^^^^^^^^^^^


```

Going back to the PR @shortdoom, I think it would be handy to preserve more of the metadata we fetch. From experience, some etherscan contracts don't build without using the same optimization/compiler flags (e.g. via-ir, or the evm target version). I think writing a suitable `foundry.toml` file would be an easy way to achieve storage and make it easily usable, although it might require adding one new `contracts` folder in the storage path, what do you all think? Is there a more generic way to register this metadata?

**Comment by shortdoom** [comment]
I dug a little deeper as some of targets were still failing for me, including `0x2a311e451491091d2a1d3c43f4f5744bdb4e773a` from `cli_tests_etherscan.sh`.

Can you try first downloading and then `cd` into the directory to re-execute compilation of Target.sol and see if it succeeds or complains about import paths?

I found few more targets like this, each having `node_modules` directory involved.

`0x19c7d0fbf906c282dedb5543d098f43dfe9f856f` also from `cli_tests_etherscan.sh`. 

However, for the above - in contrary to all of the other examples - it helps to manually provide remappings as argument `crytic-compile contracts/SNDMN.sol --solc-remaps "@openzeppelin/=node_modules/@openzeppelin/"`

And, two more, also with node_modules mapping

`0x3fC91A3afd70395Cd496C647d5a6CC9D4B2b7FAD` - fails to re-compile
`0x841AF31c35FcE829DbB939466239Dc32F96840ae` - fails to re-compile

Coming back to how to built it into the better automation I think we could just use `remappings` from here https://github.com/crytic/crytic-compile/blob/1e5deb2e06e2d606444b30767af421c9f4a2d55e/crytic_compile/platform/etherscan.py#L185-L190 and output them to the `remappings.txt` file then crytic-compile could just check for existence of this file and and compile accordingly, as @0xalpharush suggested.

**Comment by elopez** [comment]
@shortdoom I see `0x2a311e451491091d2a1d3c43f4f5744bdb4e773a` does have remappings on its metadata:

```
Remappings: ['@0x/contracts-utils=/home/cluracan/code/0x-protocol/node_modules/@0x/contracts-utils', '@0x/contracts-erc20=/home/cluracan/code/0x-protocol/contracts/zero-ex/node_modules/@0x/contracts-erc20']
Directory: crytic-export/etherscan-contracts/0x2a311e451491091d2a1d3c43f4f5744bdb4e773a-NativeOrdersFeature
Sanitized: ['@0x/contracts-utils=home/cluracan/code/0x-protocol/node_modules/@0x/contracts-utils/', '@0x/contracts-erc20=home/cluracan/code/0x-protocol/contracts/zero-ex/node_modules/@0x/contracts-erc20/']
```

It is also a good example why other metadata such as the solc version is important; even though the contracts claim `solc ^0.6.5`, they don't build with that version, but do with `0.6.12` (the one in the etherscan metadata).

Manually providing the remappings from the metadata does work [note I'm using the globbing patch from my earlier message still, to make things easier to build]

```
emilio@mbpro 0x2a311e451491091d2a1d3c43f4f5744bdb4e773a-NativeOrdersFeature % ../../../venv/bin/crytic-compile .  --solc-remaps '@0x/contracts-utils=home/cluracan/code/0x-protocol/node_modules/@0x/contracts-utils/ @0x/contracts-erc20=home/cluracan/code/0x-protocol/contracts/zero-ex/node_modules/@0x/contracts-erc20/'
INFO:CryticCompile:'solc --version' running
INFO:CryticCompile:'solc @0x/contracts-utils=home/cluracan/code/0x-protocol/node_modules/@0x/contracts-utils/ @0x/contracts-erc20=home/cluracan/code/0x-protocol/contracts/zero-ex/node_modules/@0x/contracts-erc20/ ./home/cluracan/code/0x-protocol/contracts/zero-ex/contracts/src/migrations/LibMigrate.sol --combined-json abi,ast,bin,bin-runtime,srcmap,srcmap-runtime,userdoc,devdoc,hashes,compact-format --allow-paths .,/Users/emilio/git/crytic-compile/crytic-export/etherscan-contracts/0x2a311e451491091d2a1d3c43f4f5744bdb4e773a-NativeOrdersFeature/home/cluracan/code/0x-protocol/contracts/zero-ex/contracts/src/migrations' running
INFO:CryticCompile:'solc --version' running
INFO:CryticCompile:'solc @0x/contracts-utils=home/cluracan/code/0x-protocol/node_modules/@0x/contracts-utils/ @0x/contracts-erc20=home/cluracan/code/0x-protocol/contracts/zero-ex/node_modules/@0x/contracts-erc20/ ./home/cluracan/code/0x-protocol/contracts/zero-ex/contracts/src/fixins/FixinTokenSpender.sol --combined-json abi,ast,bin,bin-runtime,srcmap,srcmap-runtime,userdoc,devdoc,hashes,compact-format --allow-paths .,/Users/emilio/git/crytic-compile/crytic-export/etherscan-contracts/0x2a311e451491091d2a1d3c43f4f5744bdb4e773a-NativeOrdersFeature/home/cluracan/code/0x-protocol/contracts/zero-ex/contracts/src/fixins' running
INFO:CryticCompile:'solc --version' running
INFO:CryticCompile:'solc @0x/contracts-utils=home/cluracan/code/0x-protocol/node_modules/@0x/contracts-utils/ @0x/contracts-erc20=home/cluracan/code/0x-protocol/contracts/zero-ex/node_modules/@0x/contracts-erc20/ ./home/cluracan/code/0x-protocol/contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol --combined-json abi,ast,bin,bin-runtime,srcmap,srcmap-runtime,userdoc,devdoc,hashes,compact-format --allow-paths .,/Users/emilio/git/crytic-compile/crytic-export/etherscan-contracts/0x2a311e451491091d2a1d3c43f4f5744bdb4e773a-NativeOrdersFeature/home/cluracan/code/0x-protocol/contracts/zero-ex/contracts/src/fixins' running
INFO:CryticCompile:'solc --version' running
INFO:CryticCompile:'solc @0x/contracts-utils=home/cluracan/code/0x-protocol/node_modules/@0x/contracts-utils/ @0x/contracts-erc20=home/cluracan/code/0x-protocol/contracts/zero-ex/node_modules/@0x/contracts-erc20/ ./home/cluracan/code/0x-protocol/contracts/zero-ex/contracts/src/fixins/FixinEIP712.sol --combined-json abi,ast,bin,bin-runtime,srcmap,srcmap-runtime,userdoc,devdoc,hashes,compact-format --allow-paths .,/Users/emilio/git/crytic-compile/crytic-export/etherscan-contracts/0x2a311e451491091d2a1d3c43f4f5744bdb4e773a-NativeOrdersFeature/home/cluracan/code/0x-protocol/contracts/zero-ex/contracts/src/fixins' running
INFO:CryticCompile:'solc --version' running
INFO:CryticCompile:'solc @0x/contracts-utils=home/cluracan/code/0x-protocol/node_modules/@0x/contracts-utils/ @0x/contracts-erc20=home/cluracan/code/0x-protocol/contracts/zero-ex/node_modules/@0x/contracts-erc20/ ./home/cluracan/code/0x-protocol/contracts/zero-ex/contracts/src/fixins/FixinCommon.sol --combined-json abi,ast,bin,bin-runtime,srcmap,srcmap-runtime,userdoc,devdoc,hashes,compact-format --allow-paths .,/Users/emilio/git/crytic-compile/crytic-export/etherscan-contracts/0x2a311e451491091d2a1d3c43f4f5744bdb4e773a-NativeOrdersFeature/home/cluracan/code/0x-protocol/contracts/zero-ex/contracts/src/fixins' running
INFO:CryticCompile:'solc --version' running
INFO:CryticCompile:'solc @0x/contracts-utils=home/cluracan/code/0x-protocol/node_modules/@0x/contracts-utils/ @0x/contracts-erc20=home/cluracan/code/0x-protocol/contracts/zero-ex/node_modules/@0x/contracts-erc20/ ./home/cluracan/code/0x-protocol/contracts/zero-ex/contracts/src/features/INativeOrdersFeature.sol --combined-json abi,ast,bin,bin-runtime,srcmap,srcmap-runtime,userdoc,devdoc,hashes,compact-format --allow-paths .,/Users/emilio/git/crytic-compile/crytic-export/etherscan-contracts/0x2a311e451491091d2a1d3c43f4f5744bdb4e773a-NativeOrdersFeature/home/cluracan/code/0x-protocol/contracts/zero-ex/contracts/src/features' running
...
```

So yeah, I think just storing the (sanitized) remappings we get from etherscan should be sufficient for remapping purposes 👍

**Comment by shortdoom** [comment]
I decided to output metadata directly to the `crytic_compile.config.json` file - that's automatically picked up by crytic-compile when re-compiling again from the downloaded directory and now no cli arguments are needed as it was the case with some sources. This is crytic-compile (solc) specific, for foundry/hardhat it would still be up to the user to build an appropriate config files.
